### DISCUSSION TITLE
pb_hooks mutating global state

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#3693] [Q&A] [answered]
I see in the docs

> keep in mind that the loaded modules use a shared registry and mutations should be avoided when possible to prevent concurrency issues

So if I have something like

```js
// globals.js
module.exports = {}
```

And then I use it via 

```
const globals = require(`${__hooks}/globals.js`)
globals.foo = 'baz'
```

What types of concurrency issues might happen? 

If it's not safe to mutate a global state, then would you please consider exposing a threadsafe map? Something like getter `globals('foo')` and setter `globals('foo','baz')`. Just a string-based store that is lost when pocketbase exits would be fine. https://github.com/dop251/goja/issues/70#issuecomment-390473530

**Comment by ganigeorgiev** [comment, marked as answer]
The goja defined variables can't cross goroutine boundaries because it is not guaranteed that they will be still available in the goja runtime stack.

For a global concurrency safe map you can use `$app.cache().set("something", 123)` (to retrieve the value call `$app.cache().get("something")`; _note that `$app.cache()` will be soft-deprecated and renamed in the future to `$app.store()`_).

**Reply by benallfree** [reply]
Great, I'll try that!
